### ISSUE TITLE
[Feature Request] Plugin profiles

### ISSUE THREAD

**Issue by tegru** [wwmm/pulseeffects#493]
Hello,

On my "Media PC" I regularly switch between TV-speakers, external speakers and headphones. Ideally I'd like to set specific plugins/settings for each device in PulseEffects. Unfortunately it's not currently practical or even possible:

- Plugins are global and must be enabled/disabled manually every time after changing output device. E.g. I'd like to use the Crossfeed plugin only with my headphones, Bass Enhancer for TV... but enabling/disabling them every time is a pain.
- The equalizer is currently only really usable with just one output device, as once you set it "perfectly" for e. g. speakers, it's wrong for your headphones and so on. This problem is elevated by the fact that it’s not possible to save your own presets.

So I'm proposing:
A feature for setting different "plugin profiles" for individual outputs/ports, which would then be applied automatically by PulseEffects when a specific output/port is selected in PulseAudio. IMHO, this would be great improvement to usability and all outputs/ports could have their own "optimized" settings.

There could also be a drop-down menu for selecting a profile manually from the main interface. Consequently, this would also add the rather handy ability to choose the output from within PulseEffects.

By default a "global profile" could be in use, so usability doesn't need to change for a casual user. Here you could also set which plugins are in use for all outputs.

_Edit: Deleted a small change..._

**Comment by wwmm** [comment]
We already have this feature. But I admit it is no obvious
![Screenshot from 2019-04-22 19-54-50](https://user-images.githubusercontent.com/16225172/56538099-870f1f80-6538-11e9-8a86-09ea85a39bae.png)
In the image above the preset with the "circular" icon pressed will be automatically loaded whenever the device currently selected as default is reconnected to the system. In your case You have to do something like this. When using the headphone press the circular icon related to the preset you want to be loaded when the headphone is plugged. Now for example set your external device as the default Pulseaudio output device and repeat the process.

**Comment by tegru** [comment]
Oh thanks. Sorry about that. Been using PulseEffects for only about a week but I still should have spotted it.

Unfortunately the automatic loading doesn't work on this setup when plugging headphones to the internal sound card. On my other computer it doesn't seem to work at all... should do more testing though.

Also, why do applications sometimes get disabled when output is changed? Is it a bug or am I doing something wrong again ;)?

**Comment by wwmm** [comment]
I think that on some soundcards the headphone is created as a new port and not as a new output device. For now this feature only works when a new output device becomes the default. What is probably happening in your case is that the default device stays the same and the headphone is just a port in it.

Try to enable the option `Process All Outputs` in the settings menu
![Screenshot from 2019-04-23 20-08-31](https://user-images.githubusercontent.com/16225172/56621635-96a76a80-6603-11e9-9ce6-4fc54574c1e8.png)
It will force all apps to be processed. If you do not do that your it will be up to Pulseaudio to remember in which device your audio application was outputting to. My guess is that its algorithm is moving the apps to the default device when you change the output.

Depending on the Linux flavor you are using it may help to take a look at our [FAQ](https://github.com/wwmm/pulseeffects/wiki/FAQ).

**Comment by wwmm** [comment]
Maybe in the future we will support port changes too. But to be honest I don't know if it is possible.

**Comment by tegru** [comment]
The `Process All Outputs` option has been enabled all the time.

Yes I believe that only a new port gets created with headphones. The irony here is that on both of my setups, there actually is a real physical "line out" output jack on the docking station, while the "headphone output" is on the laptop.

The issue with applications getting disabled is actually off-topic; It happens when I change outputs even without any presets. This is also kind of a bigger problem: change the output and find out that all your apps were silently disabled on the Applications page, and have to be re-enabled manually to get them processed again.

**Comment by wwmm** [comment]
Which program are you using to change the default device? I did some tests here on my computer and although everything is alright when using Pavucontrol to change the default device there is a problem when using Gnome control center. The audio app is disabled just like you said. It is obvious that Gnome control center isn't just changing the default device. They are iterating over all apps and moving them to this new default. I understand that for some users this may be interesting. But for anyone using custom audio routing this is terrible! Pulseaudio does not offer any programming API that would allow me to forbid Gnome control center to do this =/

I took a look at Pulseaudio programming references and I think that supporting headphone ports is not hard. With luck this week I will implement it.

**Comment by tegru** [comment]
The media pc is running Gnome where I use a shell extension to change the output device. On the other computer running Mate, I use the Sound Preferences app you get by right clicking the volume icon. The problem with apps occurs on both systems.

Good to hear that supporting ports is possible after all.

PS. The automatic loading of presets is now working also on the Mate desktop. Seems like rebooting fixed it.

**Comment by wwmm** [comment]
Is there any change if you Pavucontrol?

**Comment by tegru** [comment]
Yes there is. Looks like it works OK with Pavucontrol.

**Comment by wwmm** [comment]
Good to know. I opened an issue in Gnome gitlab page https://gitlab.gnome.org/GNOME/gnome-control-center/issues/494. Let's see what happens. As far as I can see the only thing in my reach is to add support for device ports.

**Comment by tegru** [comment]
Hmm... well the other option might be developing your own shell extension and tray volume app, which would then use the "Pavucontrol method" for changing outputs...

Actually, a tray icon is something that PulseEffects really needs anyway. It could also be used for e.g. disabling/enabling plugins, minimizing to tray and maybe even a clipping indicator... When you think of it, PulseEffects "should be" a "tray app", as it works in the background. It's not a "real app" like a browser or a text editor, and therefore should not be cluttering the desktop/taskbar all the time.

Another option is to add output selection to the main interface, like I kind of suggested in my first message. IMHO, this is something that should be there for a more complete user experience anyway. At the very least, the interface should indicate which output is currently active.

To put it another way, I'd like to see PulseEffects a complete "Audio Control Center", rather than just a interface for the effects.

I can imagine it'll take a long time before Gnome devs address this issue, if they ever will.  And how about Mate? Shouldn't the issue be reported for them too?

**Comment by Digitalone1** [comment]
There is already a Pulseaudio system tray:

https://github.com/christophgysin/pasystray
https://www.archlinux.org/packages/community/x86_64/pasystray/

I don't need to change default sink on my system, but I use this app only to check the default device from system tray and you can also switch it runtime.

**Comment by tegru** [comment]
Thanks. It works in Mate, but now I have two identical volume icons in my notification area. Looks like it's not possible to remove the original. Oh well...

**Comment by Digitalone1** [comment]
I had an identical issue on Plasma, but fortunately default pulse volume control (plasma-pa) could be uninstalled. I installed kmix (much better for me) and pasystray, then made pasystray automatically hidden and I use it only opening the system tray menu while kmix is always visible on the panel.

**Comment by tegru** [comment]
Problem solved for me too. The solution in Mate is to disable Volume Control in Startup Applications.

**Comment by wwmm** [comment]
The need for a tray icon is debatable. When users asked for it in the past what they really wanted was just to keep the effects running when PulseEffects window was closed. We are able to do that. Just enable `Start Service at Login`. And I am a little skeptical about how useful would be to enable/disable plugins in the tray icon. Personally I would never use it as in this case I would also want to take a look at the level bars, latency, etc. Maybe switching presets in the tray icon would be more interesting. In any case the main reason why this was never implemented is that there won't be tray icons in GTK4. So I would have to remove this once I start porting PulseEffects to GTK4.

I would like to keep PulseEffects focused only in the signal processing. Changing configurations like the default user audio device or the mains system volume feels like the domain of other apps. But being able to choose an output device different than the default makes sense and we already support this
![Screenshot from 2019-04-25 11-32-25](https://user-images.githubusercontent.com/16225172/56743882-cebac480-674d-11e9-8b01-a3e6a45f3f6e.png)

Assuming that Mate is not using Gnome libraries under the hood this issue should be reported to them too.

**Comment by wwmm** [comment]
As I still have to add support for device ports I suggest we keep this issue opened for a while so I do not forget to do this. I will close this issue when this feature is implemented.

**Comment by tegru** [comment]
Well here we have all the ingredients to the typical discussions regarding fragmentation on Linux and what type of desktop is better - traditional or "what ever else like Gnome". Obviously the development of GTK+ follows Gnome's direction. And so does PulseEffects, even though it doesn't even work correctly with Gnome, and maybe never will.

I'm very much old schoo

**Comment by Digitalone1** [comment]
I read it on my email. I choose to use good applications, I don't mind if they're from Plasma or Gnome.

My tastes are that Qt applications look better, but I can use GTK applications if they're better. For example, there are some Qt browsers, but I use Chromium and Firefox.

And even if I'm on Plasma that refuses CSD, I like them on certain apps. I'm using Pulseeffect and GTK browsers with CSD.

**Comment by wwmm** [comment]
No problems. I used KDE for many years and at that time I used QT+QML in my small projects. But eventually I started to feel more comfortable with GTK3 and GNOME approach. This doesn't mean I like everything in their philosophy. Removing useful features just to make an interface cleaner is not my style. If it is useful it will be there. But in the case of the tray icons GTK4 doesn't give me a choice. 

And thinking about past issues we had our share of problems with KDE and even with some Pulseaudio's modules. I think that the reason is that no application tried to do in the past the kind of things that PulseEffects does. So people took design decisions without considering the side effects to custom audio routing.

**Comment by wwmm** [comment]
Master branch code updated. Headphones should trigger the preset auto load now.
